### ISSUE TITLE
If an object has multiple IFCSHAPEREPRESENTATION geometries in the same context, it would be nice if BB rendered them out.

### ISSUE THREAD

**Issue by theoryshaw** [IfcOpenShell/IfcOpenShell#5877]
If an object has multiple IFCSHAPEREPRESENTATION geometries in the same context, it would be nice if BB rendered them out at the same time.  It seems other viewers do. 

[multiple representation items - Copy.ifc.txt](https://github.com/user-attachments/files/18137928/multiple.representation.items.-.Copy.ifc.txt)

![20241214_1552](https://github.com/user-attachments/assets/0b431314-1352-4acf-9377-c5531b79dc39)

![image](https://github.com/user-attachments/assets/192f2d0d-0863-4c23-a51b-09c0a85bd70d)
![20241214_1552](https://github.com/user-attachments/assets/7b3b244e-9663-4cc3-ba2a-7da76961051e)

![image](https://github.com/user-attachments/assets/998afc31-a88a-4f1d-b626-77e0cb95b911)

**Comment by theoryshaw** [comment]
somewhat related: https://github.com/IfcOpenShell/IfcOpenShell/issues/3240

**Comment by Andrej730** [comment]
Related to https://github.com/buildingSMART/IFC4.3.x-development/issues/193

From the specs it's not prohibited and ambiguous but it kind of meant to be "each representation must be a valid representation in the context on it's own". (Maybe I'm reading it that way because I want it to be 1 representation to cover it as it's much simpler 😅).

https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcProductRepresentation.htm:
> Contained list of representations (including shape representations). Each member defines a valid representation of a particular type within a particular representation context.

**Comment by aothms** [comment]
We're definitely moving into that direction. See for example:

https://github.com/buildingSMART/ifc-gherkin-rules/blob/main/features/GEM003_Unique-representation-identifier.feature

**Comment by theoryshaw** [comment]
Is there currently an approach in the IFC schema that accommodates an ifcproduct with multiple RepresentationTypes? 

For example, can you have a door product that uses a IfcExtrudedAreaSolid(SweptSolid) for the door panel, an IfcSphere(CSG) for the door knob, an IfcIndexedPolygonalFace(Tessellation) for the door knocker and an IfcSectionedSolidHorizontal(AdvancedSweptSolid) for the trim?

**Comment by Andrej730** [comment]
I don't think there is. In case of a need of mixed representation types, IFC only offers using some more general representation type like IfcIndexedPolygonalFace or creating a weird aggregate that will combine items of different types.

**Comment by aothms** [comment]
> a weird aggregate 

I know it has downsides, but imho this still makes the most sense. If it has a different item type, it would most likely also carry different properties.

**Comment by theoryshaw** [comment]
If this is the general direction, how does the schema accomodate an aggregate that has a one-to-many relationship?  If this was possible, the 'downsides' of not allowing a type to have mixed representation types could be more palatable.

**Comment by Moult** [comment]
I'm going to make a call here that it's basically unsupported especially given @TLiebich 's blessing upstream three years ago.

**Comment by aothms** [comment]
Agreed. If they have the same identifier then the validation service will
also flag it as an error.

Sent from a mobile device, excuse my brevity. Kind regards, Thomas

Op za 15 feb 2025 08:49 schreef Dion Moult ***@***.***>:

**Comment by theoryshaw** [comment]
Had an interesting chat...
https://claude.ai/share/4963d195-ed14-4b82-a7c9-20c967d7e470

**Comment by aothms** [comment]
I see three options:

- Accept as is and suffer from the pain
- Introduce some sort of IfcPatch recipe that takes TypeObject-level decompition and duplicates them to the Occurences, and vice versa removes them. Somehow support this in Bonsai, but also Blender can probably not natively render instanced sub-object trees.
- Move to IFC5 where all of this is happily supported in the spec, but an efficient Copy-on-write mechanism still needs to be implemented so that inherited sub-object trees are not duplicated in memory.

There's no happy paths here, but it would be good if in the community we can find some closure in a direction.

**Comment by theoryshaw** [comment]
Thank you, just throwing this out there, could there be a fourth option hiding between 2 and 3?

What if instead of full copy-on-write (which is genuinely hard to implement correctly), we just started with occurrences inheriting their type's decomposition as-is, no overrides? If a specific occurrence needs to differ, it gets its own type. Simple rule, no merging logic, no virtual instance management.

The nice thing is Blender already understands this concept natively... it's basically collection instancing. And it solves the file bloat problem completely without waiting for IFC5.

The obvious tradeoff is you'd get more types. But honestly... if door 47 really is different, shouldn't it be a different type anyway?

Would it be worth treating this as an explicit interim position rather than jumping straight to the full IFC5 mechanism?

Simply put, as an author, i would get a ton of mileage out of strict/immutable inheritance alone.
